src/main/java/edu/harvard/iq/dataverse/api/Datasets.java

```diff
@@ -2429,4 +2442,345 @@ public Response getTimestamps(@PathParam("identifier") String id) {
+                        }
+
+                        msg("ADD!  = " + newFilename);
+
+                        //-------------------
+                        // Run "runAddFileByDatasetId"
+                        //-------------------
+
+                        addFileHelper.runAddFileByDataset(dataset,
+                                newFilename,
+                                newFileContentType,
+                                newStorageIdentifier,
+                                null,
+                                optionalFileParams, true);
+
+                        if (addFileHelper.hasError()) {
+
+                            JsonObjectBuilder fileoutput = Json.createObjectBuilder()
+                                    .add("storageIdentifier ", newStorageIdentifier)
+                                    .add("error Code: ", addFileHelper.getHttpErrorCode().toString())
+                                    .add("message ", addFileHelper.getErrorMessagesAsString("\n"));
+
+                            jarr.add(fileoutput);
+
+                        } else {
+                            String successMsg = BundleUtil.getStringFromBundle("file.addreplace.success.add");
+
+                            JsonObject successresult = addFileHelper.getSuccessResultAsJsonObjectBuilder().build();
+
+                            try {
+                                logger.fine("successMsg: " + successMsg);
+                                String duplicateWarning = addFileHelper.getDuplicateFileWarning();
+                                if (duplicateWarning != null && !duplicateWarning.isEmpty()) {
+                                    // return ok(addFileHelper.getDuplicateFileWarning(), addFileHelper.getSuccessResultAsJsonObjectBuilder());
+                                    JsonObjectBuilder fileoutput = Json.createObjectBuilder()
+                                            .add("storageIdentifier ", newStorageIdentifier)
+                                            .add("warning message: ", addFileHelper.getDuplicateFileWarning())
+                                            .add("message ", successresult.getJsonArray("files").getJsonObject(0));
+                                    jarr.add(fileoutput);
+
+                                } else {
+                                    JsonObjectBuilder fileoutput = Json.createObjectBuilder()
+                                            .add("storageIdentifier ", newStorageIdentifier)
+                                            .add("message ", successresult.getJsonArray("files").getJsonObject(0));
+                                    jarr.add(fileoutput);
+                                }
+
+                            } catch (Exception ex) {
+                                Logger.getLogger(Files.class.getName()).log(Level.SEVERE, null, ex);
+                                return error(Response.Status.BAD_REQUEST, "NoFileException!  Serious Error! See administrator!");
```

Comment:
If these are Serious Exceptions, are they 500 errors rather than Bad Requests?
Note that one option these days is to, rather than catching exceptions in the api call is to let the .api.errorhandler classes create responses. For example - the json parse errors above could just be thrown and the JsonParseExceptionHandler class would log and send the right response. Similarly, if there are NullPointerExceptions or other internal/runtime errors, the ThrowableHandler will turn them into 500 errors.